c2rust-analyze/src/c_void_casts.rs

```diff
@@ -158,25 +162,33 @@ pub struct CVoidCast<'tcx> {
 /// to their properly typed pointers,
 /// like [`CVoidCasts`], but in a single direction, meaning it represents either
 /// [`CVoidCastDirection::From`] or [`CVoidCastDirection::To`].
-#[derive(Default)]
-pub struct CVoidCastsUniDirectional<'tcx>(HashMap<CVoidPtr<'tcx>, Place<'tcx>>);
+#[derive(Default, Clone, Debug)]
+pub struct CVoidCastsUniDirectional<'tcx>(HashMap<Location, (CVoidPtr<'tcx>, Place<'tcx>)>);
 
 impl<'tcx> CVoidCastsUniDirectional<'tcx> {
-    pub fn contains(&self, place: Place<'tcx>) -> bool {
-        self.0.contains_key(&place)
+    pub fn contains(&self, loc: &Location) -> bool {
+        self.0.contains_key(loc)
     }
 
     /// Get the adjusted [`Place`], skipping over [`*c_void`](core::ffi::c_void) intermediaries.
     ///
     /// That is, if `place` is a [`CVoidPtr`] in this map of [`CVoidCast`]s,
     /// then the [`Place`] of its other, property-typed pointer is returned.
     /// Otherwise, the same `place` is returned, as no adjustments are necessary.
-    pub fn get_adjusted_place_or_default_to(&self, place: Place<'tcx>) -> Place<'tcx> {
-        *self.0.get(&place).unwrap_or(&place)
+    pub fn get_adjusted_place_or_default_to(
+        &self,
+        loc: &Location,
+        place: Place<'tcx>,
+    ) -> Place<'tcx> {
+        *self
+            .0
+            .get(loc)
+            .map(|(_void, subst)| subst)
+            .unwrap_or(&place)
     }
 
-    pub fn insert(&mut self, cast: CVoidCast<'tcx>) {
-        self.0.insert(cast.c_void_ptr, cast.other_ptr);
+    pub fn insert(&mut self, loc: Location, cast: CVoidCast<'tcx>) {
+        self.0.insert(loc, (cast.c_void_ptr, cast.other_ptr));
```

Comment:
This is maybe excessively cautious, but could you assert here than `loc` isn't already in the map?  This should always be the case since we currently have at most one "from" cast and at most one "to" cast per `Location`.

Reply:
91bad4f2